pkg/redirectpolicy/redirectpolicy.go

```diff
@@ -0,0 +1,235 @@
+//  Copyright 2020 Authors of Cilium
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package redirectpolicy
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/cilium/cilium/pkg/k8s"
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
+	k8sUtils "github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/policy/api"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type lrpConfigType = int
+
+const (
+	lrpConfigTypeNone = iota
+	lrpConfigTypeAddr
+	lrpConfigTypeSvc
+)
+
+// LRPConfig is the internal representation of Cilium Local Redirect Policy.
+type LRPConfig struct {
+	name      string
+	namespace string
+	id        policyID
+	uid       types.UID
+	lrpType   lrpConfigType
+	// frontendMappingForAddr stores a mapping of policy config's address matcher
+	// frontend to a list of its associated backends
+	frontendMappingForAddr *feMapping
+	// frontendMappingForSvc is a map with policy config's service port name as
+	// the key and mapping between the service frontend and list of its
+	// associated backends matched by the port name as the value.
+	frontendMappingForSvc map[fePortName]*feMapping
```

Comment:
I'm a little concerned that these fields represent "realized state" (mutable, changed based on actions taken in the manager) while the rest of this structure is "desired state"(immutable, declared by the user). I wonder whether they should actually hang off the `manager` directly via a map using the LRP's ID. I don't see them as representing part of the user-specified "LRP config" per the structure name.

For context, I'm cross-referencing their use against `rpm.upsertServiceWithBackends()` which updates these fields just before pushing the upsert events into the service manager.

Even if we think it makes sense for them to remain in this structure, I think it would be good to put them towards the end with a clear separation and comment that describes the difference in access patterns to hopefully avoid any misunderstanding in their meaning.

Reply:
Upon subsequent reading, I think that my initial read on these fields was incomplete. It's certainly still mutable state, but it seems more like the backends in these fields represent the calculated backends that currently should be configured, which would be a cache of the desired state. Then if the user deletes specific backends for instance, the code subtracts the "to-delete" backends from these cached understanding of datapath state to end up with the new calculated set of backends. This would be rather than reaching down into the datapath to figure out the diff. The access patterns are certainly different, and they're mutable rather than immutable, so there might still be an argument for placing them outside the LRPConfig, but I'm not as concerned as with the previous read through.